### PR TITLE
Credit payment request with token

### PR DESCRIPTION
**Pull request by hamadata**
This feature branch implements credit card payment process with token.

[motivation]
- authorization, create_customer and update_customer should be processed without sending credit card information to web application server if servers are not PCI-DSS compliant.

[other]
- In sales.rb, :processing_datetime has default value.

### CODE REVIEW

**Comment by hamadata** [comment]
@kyanny 
I appreciate if this pull request is reviewed, but not required for now.
I just would like to let developers know about this branch and want to discuss if any improvements are required.

**Comment by kyanny** [comment]
Thank you very much @hamadata !!
We will look into it cc @banyan @kachick

**Comment by hamadata** [comment]
This features is now working in my project.
I believe this branch can be merged.

File: README.md [hunk 1]
```diff
@@ -55,20 +55,23 @@ Currently client side encoding is supported only for UTF-8. (UTF-8 to Shift-JIS)
...
-返金要求            | `ST02-00303-101`   | `Sbpayment::API::Credit::RefundRequest`               |
-部分返金要求        | `ST02-00307-101`   | `Sbpayment::API::Credit::PartlyRefundRequest`         |
-顧客情報登録        | `MG02-00101-101`   | `Sbpayment::API::Credit::CreateCustomerRequest`       |
-顧客情報更新        | `MG02-00102-101`   | `Sbpayment::API::Credit::UpdateCustomerRequest`       |
-顧客情報削除        | `MG02-00103-101`   | `Sbpayment::API::Credit::DeleteCustomerRequest`       |
-顧客情報参照        | `MG02-00104-101`   | `Sbpayment::API::Credit::ReadCustomerRequest`         |
+Action Name                   | sps-api-request id | Request class                                         |
+---                           | ---                | ---                                                   |
+与信要求                      | `ST01-00111-101`   | `Sbpayment::API::Credit::AuthorizationRequest`        |
+与信要求(トークン)            | `ST01-00131-101`   | `Sbpayment::API::Credit::AuthorizationRequestToken`   |
```
Comment (kachick) [inline README.md]: s/AuthorizationRequestToken/AuthorizationTokenRequest/ ?

File: README.md [hunk 2]
```diff
@@ -55,20 +55,23 @@ Currently client side encoding is supported only for UTF-8. (UTF-8 to Shift-JIS)
...
+再与信要求                    | `ST01-00113-101`   | `Sbpayment::API::Credit::ReAuthorizationRequest`      |
+コミット (確定処理)           | `ST02-00101-101`   | `Sbpayment::API::Credit::CommitRequest`               |
+売り上げ連携                  | `ST02-00201-101`   | `Sbpayment::API::Credit::SalesRequest`                |
+決済結果参照                  | `MG01-00101-101`   | `Sbpayment::API::Credit::InquireAuthorizationRequest` |
+取消要求                      | `ST02-00305-101`   | `Sbpayment::API::Credit::CancelAuthorizationRequest`  |
+返金要求                      | `ST02-00303-101`   | `Sbpayment::API::Credit::RefundRequest`               |
+部分返金要求                  | `ST02-00307-101`   | `Sbpayment::API::Credit::PartlyRefundRequest`         |
+顧客情報登録                  | `MG02-00101-101`   | `Sbpayment::API::Credit::CreateCustomerRequest`       |
+顧客情報更新                  | `MG02-00102-101`   | `Sbpayment::API::Credit::UpdateCustomerRequest`       |
+顧客情報登録(トークン)        | `MG02-00131-101`   | `Sbpayment::API::Credit::CreateCustomerRequestToken`  |
```
Comment (kachick) [inline README.md]: `s/RequestToken/TokenRequest/` ?

File: README.md [hunk 3]
```diff
@@ -55,20 +55,23 @@ Currently client side encoding is supported only for UTF-8. (UTF-8 to Shift-JIS)
...
+顧客情報更新(トークン)        | `MG02-00132-101`   | `Sbpayment::API::Credit::UpdateCustomerRequestToken`  |
```
Comment (kachick) [inline README.md]: `s/RequestToken/TokenRequest/` ?

File: lib/sbpayment/api/credit/authorization_token.rb [hunk 4]
```diff
@@ -0,0 +1,66 @@
...
+          key :divide_times,  encrypt: true
+        end
+        class PayOptionManage
+          include ParameterDefinition
+
+          tag 'pay_option_manage'
+          key :token
+          key :token_key
+          key :cust_manage_flg
+          key :cardbrand_return_flg, default: '1'
```
Comment (kachick) [inline lib/sbpayment/api/credit/authorization_token.rb]: Hmm... The documents looks `0` to me. But we already implemented some values with difference.
I don't have confident for the origin policy. @banyan Could you review?

**Comment by kachick** [comment]
📝 Note to me.

* https://github.com/quipper/sbpayment.rb/pull/50/files?w=1 makes better diff in this PR.
* Related SBPS documents are below.
    * ST01-00131-101: `4OP_Y_システム仕様書（リアル与信 トークン利用）1.0.pdf`
      * 

    * MG02-00131-101: `4OP_Y_システム仕様書（顧客情報 登録 トークン利用）1.0.pdf`
    * MG02-00132-101: `4OP_Y_システム仕様書（顧客情報 更新 トークン利用）1.0.pdf`
      *

File: lib/sbpayment/api/credit/authorization_token.rb (hunk 4 above)
Comment (kachick) [inline lib/sbpayment/api/credit/authorization_token.rb]: It is not an issue in this PR, we can leave for now!

**Comment by kachick** [comment]
@hamadata Thanks for your appreciate! 😂 

Could you check some typos (or not)? ref: https://github.com/quipper/sbpayment.rb/pull/50#discussion_r97016486

**Comment by hamadata** [comment]
@kachick 
Thank you for your review!

As you have suggested, I agree, e.g. AuthorizationTokenRequest is more natural than AuthorizationRequestToken.

But in my opinion, for instance, the behavior of AuthorizationTokenRequest is similar to AuthorizationRequest, so I add suffix "Token" to non-token request file name. Everyone can easily recognize each set of files behaves almost the same.
 
What do you feel about the idea? (Just a opinion. I will follow your conclusion.)

File: lib/sbpayment/api/credit/authorization_token.rb [hunk 5]
```diff
@@ -0,0 +1,66 @@
+require_relative '../../request'
+require_relative '../../response'
+
+module Sbpayment
+  module API
+    module Credit
+      class AuthorizationTokenRequest < Request
```
Comment (kachick) [inline lib/sbpayment/api/credit/authorization_token.rb]: Or typo is this? 🤔 https://github.com/quipper/sbpayment.rb/pull/50/files#r97016486

**Comment by kachick** [comment]
Sorry to misunderstood if I am wrong, it looks just a typo to me 🙇 

Because the code already defined as `AuthorizationTokenRequest` at https://github.com/quipper/sbpayment.rb/pull/50/files#diff-d261ee7666256bf976546e0949a58506R7. But the document says `AuthorizationRequestToken` at https://github.com/quipper/sbpayment.rb/pull/50/files#diff-04c6e90faac2675aa89e2176d2eec7d8R61.

Hmm... 🤔

File: lib/sbpayment/api/credit/authorization_token.rb (hunk 5 above)
Comment (hamadata) [inline lib/sbpayment/api/credit/authorization_token.rb]: Sorry, it is my mistake 😢 
I want to unify the class names, which do you prefer, ~RequestToken or ~TokenRequest?
(If you do not care about it, I would like to fix these class names to ~TokenRequest)

Comment (kachick) [inline lib/sbpayment/api/credit/authorization_token.rb]: Ah no problem!

So... In my personal,  `~TokenRequest` sounds good to me 😄

**Comment by hamadata** [comment]
Thanks, I fixed the class names on README.md.

**Comment by kachick** [comment]
Thanks for your all work! 🎉

**Comment by kachick** [comment]
`v0.9.0` released 🎉

* https://rubygems.org/gems/sbpayment/versions/0.9.0
* https://github.com/quipper/sbpayment.rb/releases/tag/v0.9.0

**Pull request merged**